src/impy/kinematics.py

```diff
@@ -99,164 +33,77 @@ class CompositeTarget(object):
-            == len(self.component_Z)
-            == len(self.component_fractions)
-            == len(self._component_orig_fractions)
-        ):
-            raise RuntimeError(
-                "CompositeTarget:_normalize len of arrays should be equal"
-            )
+        fractions = np.empty(len(components))
+        self._components = []
+        for i, (particle, amount) in enumerate(components):
+            fractions[i] = amount
+            self._components.append(_normalize_particle(particle))
+        self._fractions = fractions / np.sum(fractions)
 
-        if not (np.sum(self.component_fractions) == 1.0):
-            raise RuntimeError("Normalization failed")
-
-    def add_component(self, az, fraction, name=""):
-        """Add material for composite target.
-
-        Fraction needs relative specification, in percent, number,
-        fraction of one, etc. Just make sure it's the same definition
-        for all components. Internal list is sorted according to
-        relative fractions.
-        """
-        self._add_component(az, fraction, name)
-        self._normalize()
-
-    def _sort(self):
-        """Sorts list acording to fraction"""
+    @property
+    def fractions(self):
+        return self._fractions
 
-        def sort_list(k, idcs):
-            return [k[i] for i in idcs]
+    @property
+    def components(self):
+        return self._components
 
-        idcs = np.argsort(self.component_fractions)
-        self.component_fractions = self.component_fractions[idcs]
-        self._component_orig_fractions = sort_list(self._component_orig_fractions, idcs)
-        self.component_A = sort_list(self.component_A, idcs)
-        self.component_Z = sort_list(self.component_Z, idcs)
-        self.component_name = sort_list(self.component_name, idcs)
+    @property
+    def Z(self):
+        """Return maximum charge number."""
+        # needed for compatibility with PDGID interface and for dpmjet initialization
+        return max(p.Z for p in self._components)
 
-    def _get_maximum_AZ(self):
-        a_val = self.component_A
-        max_ind = a_val.index(max(a_val))
-        return self.component_A[max_ind], self.component_Z[max_ind]
+    @property
+    def A(self):
+        """Return maximum number of nucleons."""
+        # needed for compatibility with PDGID interface and for dpmjet initialization
+        return max(p.A for p in self._components)
 
-    def _get_random_AZ(self):
-        """Return randomly an (A, Z) tuple according to the component fraction."""
-        ic = self.rng.choice(self.ncomponents, 1, p=self.component_fractions)[0]
-        return self.component_A[ic], self.component_Z[ic]
+    @property
+    def is_nucleus(self):
+        return True
 
-    def __str__(self):
-        ostr = "Composite target '" + self.label + "' \n"
+    def __int__(self):
+        """Return PDGID for heaviest of elements."""
+        return max(int(m) for m in self._components)
```

Comment:
So, the higher is int(m), the larger is mass? I just cannot understand connection between mass and pdg.

Reply:
It is now fixed.